waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/web/controllers/exchange/ExchangeUtils.kt

```diff
@@ -1,16 +1,115 @@
+                    ) {
+                        "id property of the first entry of the authentication relationship of the resolved did document is not a string"
+                    }
+                    firstAuthRelEntry["id"]!!.jsonPrimitive.content
+                }
+
+                is JsonPrimitive -> {
+                    check(firstAuthRelEntry.isString) { "first entry of the authentication relationship of the resolved did document is encoded as a reference to a verification method but is not of type string" }
+                    firstAuthRelEntry.content
+                }
+
+                else -> {
+                    throw IllegalArgumentException("resolved did document's authentication relationship first entry is neither a json object nor a json primitive of type string")
+                }
+            }
         }
+    }.onFailure {
+        throw it
     }
+
+    fun getW3cJwtVpProofParametersFromWalletCredentials(
+        did: String,
+        didAuthKeyId: String,
+        presentationId: String,
+        audience: String,
+        nonce: String?,
+        credentials: List<WalletCredential>,
+        disclosures: Map<String, List<String>>?,
+    ) = CredentialFilterUtils.getJwtVcList(
+        credentials,
+        disclosures,
+    ).takeIf { it.isNotEmpty() }?.let { jwtVcList ->
+        W3cJwtVpProofParameters(
+            header = mapOf(
+                "kid" to didAuthKeyId.toJsonElement(),
+                "typ" to "JWT".toJsonElement(),
+            ),
+            payload = mapOf(
+                "sub" to did.toJsonElement(),
+                "nbf" to Clock.System.now().minus(1.minutes).epochSeconds.toJsonElement(),
+                "iat" to Clock.System.now().epochSeconds.toJsonElement(),
+                "jti" to presentationId.toJsonElement(),
+                "iss" to did.toJsonElement(),
+                "aud" to audience.toJsonElement(),
+                "nonce" to (nonce ?: "").toJsonElement(),
+                "vp" to mapOf(
+                    "@context" to listOf("https://www.w3.org/2018/credentials/v1").toJsonElement(),
+                    "type" to listOf("VerifiablePresentation").toJsonElement(),
+                    "id" to presentationId.toJsonElement(),
+                    "holder" to did.toJsonElement(),
+                    "verifiableCredential" to jwtVcList.toJsonElement()
+                ).toJsonElement(),
+            ).filterValues { it.toString().isNotBlank() },
+        )
+    }
+
+    fun getIETFJwtVpProofParametersFromWalletCredentials(
+        keyId: String,
+        audience: String,
+        nonce: String?,
+        credentials: List<WalletCredential>,
+        disclosures: Map<String, List<String>>?,
+    ) = credentials.filter { it.format == CredentialFormat.sd_jwt_vc }.map { credential ->
+        val serializedVcWithDisclosures = listOf(
+            credential.document
+        ).plus(
+            if (disclosures?.containsKey(credential.id) == true) {
```

Comment:
maybe this could be a separate method, e.g. getDisclosures:
```kotlin
private fun getDisclosures(
    disclosures: Map<String, List<String>>?,
    credentialId: String
) = if (disclosures?.containsKey(credentialId) == true) {
    "~${disclosures[credentialId]!!.joinToString("~")}~"
} else "~"
```

Reply:
fixed in f489dda6.